### ISSUE TITLE
Option to disable Id Inference

### ISSUE THREAD

**Issue by kendallmiller** [elastic/elasticsearch-net#3578]
We're working on heavily optimizing our data ingest process to connect Loupe to Elasticsearch for our customers.  One problem we've run into is that we're getting a performance hit because virtually every object we have has a property called "Id" which is being automapped by NEST to be a document Id in Elastic and in turn meaning Elastic is evaluating these for uniqueness as we insert.

Don't get me wrong, in a world where all things are free we wouldn't care but a bit of experimentation has shown that this may be significantly slowing down ingestion rates when you get to the 30-50 million document size.  We'd like to disable this because we just don't need or want Elastic to treat this like a key but we haven't been able to come up with any way that allows us to have this property on our documents but *not* get it auto-mapped to __id.  

You could argue that we could make new objects that have this field renamed to bypass the mapper but we're all about that performance and taking objects and recreating new objects to put them in ES, then getting them out of ES and making new objects to convert them back, feels like a *lot* of burn and churn to bypass some automagic in the .NET client.

What we'd like is an option to disable the auto inference of this - ideally either on the .Settings() of CreateIndexDescriptor or the Map<T>() function.

**Comment by russcam** [comment]
Thanks for opening @kendallmiller. We'll look to discuss this internally next week

**Comment by kendallmiller** [comment]
I did get around this in a slightly ridiculous way, added here for others that are pondering how to get around this - I created a custom DTO just for indexing the data and on that renamed the Id property so it wouldn't get auto-mapped.  I then added an attribute to tell Elasticsearch to index that property as "id".  This seems to outsmart the inference routine so it doesn't map the property to __id but does record it in the source as id.  Then, when I query the data back into the original DTO the Id property is correctly there.

In our case this does mean we're making an extra copy of each object between our original data and serializing it but in the grand scheme of the work done to index the document this is not measurable in either GC pressure or overall memory utilization, probably because we're already using a streaming approach so the intervening original DTO doesn't live very long.

```cs
        /// <summary>
        /// The globally unique id of this sample
        /// </summary>
        [Keyword(Name="id")]
        public Guid OriginalId { get; set; }
```

**Comment by russcam** [comment]
#3589 has been merged and will be in the next release, 6.6.0.

With this change, it will be possible to disable Id inference for all types globally, on `ConnectionSettings`:

```c#
var settings = new ConnectionSettings()
    .DefaultDisableIdInference();

var client = new ElasticClient(settings);
```

or per individual type basis:

```c#
var settings = new ConnectionSettings()
    .DefaultMappingFor<MyPoco>(m => m
        .DisableIdInference()
    );

var client = new ElasticClient(settings);
```
Note that disabling inference globally cannot be overridden on a per type basis.

I'm going to close this issue.
